### PR TITLE
Initial Python 3 port of the Python SDK

### PR DESCRIPTION
**Pull request by swgillespie**
I have learned a lot about Python 3 the past few days. This PR description will be kinda long but I hope it frames some of the challenges  of supporting Python 2 and Python 3 in the same codebase.

This PR passes all tests in `pulumi/pulumi`, but there are not many integration tests using Python. We'll need to improve these tests.

There are three commits in this PR, two of which are machine-generated:

## https://github.com/pulumi/pulumi/commit/49ef1fbd5afe1c082d6e2477ed2539ea713a2b31
This commit is a machine-assisted Python 2 to Python 3 translation, using the [`futurize` tool](http://python-future.org/automatic_conversion.html). The sorts of fixes that it made fall into three main categories:

### Relative imports (https://github.com/pulumi/pulumi/issues/1541)
The PEP that introduced this break is [here](https://www.python.org/dev/peps/pep-0328/). The basic rationale behind the break is that, in Python 2, it is ambiguous what this statement will do:

```python
import sys
```

The ambiguity is about what `foo` refers to - is this a relative import of a module `sys.py`, sitting adjacent to the current file? Or is this an absolute import of the `sys` module in the standard library?

To address this, Python 3 considers `import <identifier>` to unambiguously be an absolute import. In order to do a relative import, we must instead use `from . import <identifier>` to import a module adjacent to the module performing the import. We can opt-in to this behavior in Python 2 using `from __future__ import absolute_import`.

This commit changes all ambiguously absolute imports into relative imports and turns on `from __future__ import absolute_import`

### Old-style classes

Python 2 has both "old-style" and "new-style" classes, which differ in [esoteric ways](https://www.python.org/download/releases/2.2.3/descrintro/), while Python 3 has only "new-style" classes. New-style classes in Python 2 must derive from `object`, while all classes derive from `object` in Python 3 by default. This commit adds `object` as the explicit superclass of all classes that don't declare it.

### Collection iterators and `list`

Python 3 changed some common collection iteration methods to return iterators instead of lists. For uniformity, the automated tool converted all uses of these methods to wrap their results in `list`, which ensures uniform behavior across Python versions.

### The `str` and `basestring` types (https://github.com/pulumi/pulumi/issues/1540)

Python 2 and 3 differ significantly on string representation. To ensure uniformity across releases, this commit tests strings against `six.string_types` instead of `basestring` for type checks. The `Unknown` class also derives from `str` to get the right behavior across versions.

## https://github.com/pulumi/pulumi/commit/30fa7ba964e30b369fca7b2eb8da969c81bb35e7

This commit works around an unfortunate `protoc` bug (https://github.com/google/protobuf/issues/1491). The problem is that, if your proto files are in the same top-level directory and one proto file imports the other, `protoc` generates an ambiguously absolute import for adjacent proto modules which doesn't work with Python 3. The official workaround is to change the directory structure containing your proto files, but this is troublesome for our build system. Like many people before me, I reached to `sed` to fix this problem. It's not elegant, but it works.

## https://github.com/pulumi/pulumi/commit/982b3f6ac0a22a2b92c2819c25777d162eeed9aa

This commit contains all of the re-generated protobuf and gRPC files.

### CODE REVIEW

**Comment by swgillespie** [comment]
Not sure about reviewers but I'd appreciate any feedback on this code! Most of it is autogenerated but it does get us to a point where we can run pulumi/pulumi tests using Python 3.

File: sdk/proto/generate.sh [hunk 1]
```diff
@@ -40,6 +40,13 @@ $PROTOC --js_out=$JS_PROTOFLAGS:$JS_PULUMIRPC --grpc_out=$JS_PULUMIRPC --plugin=
 PY_PULUMIRPC=../python/lib/pulumi/runtime/proto/
 echo -e "\tPython: $PY_PULUMIRPC"
 mkdir -p $PY_PULUMIRPC
-python -m grpc_tools.protoc -I./ --python_out=$PY_PULUMIRPC --grpc_python_out=$PY_PULUMIRPC *.proto
+
+
+TEMP_DIR=$(mktemp -d)
```
Comment (mmdriley) [inline sdk/proto/generate.sh]: let's use an `EXIT` trap, like: https://github.com/pulumi/pulumi/blob/776c538b7f53295d65d88a0e4e4a8e93c1f8ac53/dist/install.sh#L23

File: sdk/proto/generate.sh [hunk 2]
```diff
@@ -40,6 +40,13 @@ $PROTOC --js_out=$JS_PROTOFLAGS:$JS_PULUMIRPC --grpc_out=$JS_PULUMIRPC --plugin=
 PY_PULUMIRPC=../python/lib/pulumi/runtime/proto/
 echo -e "\tPython: $PY_PULUMIRPC"
 mkdir -p $PY_PULUMIRPC
-python -m grpc_tools.protoc -I./ --python_out=$PY_PULUMIRPC --grpc_python_out=$PY_PULUMIRPC *.proto
+
+
```
Comment (mmdriley) [inline sdk/proto/generate.sh]: it'll be useful to have a comment about what's going on here

File: sdk/proto/generate.sh [hunk 3]
```diff
@@ -40,6 +40,13 @@ $PROTOC --js_out=$JS_PROTOFLAGS:$JS_PULUMIRPC --grpc_out=$JS_PULUMIRPC --plugin=
...
+TEMP_DIR=$(mktemp -d)
+echo -e "\tPython temp dir: $TEMP_DIR"
+python -m grpc_tools.protoc -I./ --python_out=$TEMP_DIR --grpc_python_out=$TEMP_DIR *.proto
```
Comment (mmdriley) [inline sdk/proto/generate.sh]: Always quote shell variables and glob expansions:
`"$TEMP_DIR"`
`"*.proto"`

File: sdk/proto/go/provider.pb.go [hunk 4]
```diff
@@ -894,7 +893,7 @@ func init() { proto.RegisterFile("provider.proto", fileDescriptor5) }
 
 var fileDescriptor5 = []byte{
 	// 840 bytes of a gzipped FileDescriptorProto
-	0x1f, 0x8b, 0x08, 0x00, 0x00, 0x09, 0x6e, 0x88, 0x02, 0xff, 0xc4, 0x56, 0xdd, 0x6e, 0xc3, 0x34,
+	0x1f, 0x8b, 0x08, 0x00, 0x00, 0x00, 0x00, 0x00, 0x02, 0xff, 0xc4, 0x56, 0xdd, 0x6e, 0xc3, 0x34,
```
Comment (mmdriley) [inline sdk/proto/go/provider.pb.go]: any idea why this representation is changing?

**Comment by swgillespie** [comment]
> this may be ugly, but: any way for us to ensure in CI that we don't allow Python 2/3 incompatibilities to creep in later? Can we run futurize as a linter?

We can't, unfortunately. I read a bunch this weekend about how people do this (Python 2 and Python 3 co-existing in the same codebase), and the general consensus seems to be that

1. There is no substitute for tests, nothing matters if you're not running your tests on both Python 2 and Python 3. Today I'm trying to bring up Python language host tests (like we already have for Node) so that we can be very aggressive about our test coverage without requiring us to resort to integration tests.

2. Crank up the intensity of `pylint`'s compatibility lints. They're off by default because Pylint assumes that you don't care about Python 2 and Python 3 co-existing in the same project. I'm also planning on doing this.

File: sdk/proto/go/provider.pb.go (hunk 4 above)
Comment (swgillespie) [reply sdk/proto/go/provider.pb.go]: No idea. I'll check it out.

File: sdk/proto/generate.sh (hunk 3 above)
Comment (swgillespie) [reply sdk/proto/generate.sh]: If I'm globbing, I can't quote, right? I'm using the shell to expand the glob and not protoc.

Comment (mmdriley) [reply sdk/proto/generate.sh]: Yes. That latter advice was misguided -- sorry.

```
$ touch "a.txt" "z.txt" "this name has spaces in it.txt"
$ bash -c 'echo $0; echo $1; echo $2' ./*.txt
./a.txt
./this name has spaces in it.txt
./z.txt
```

File: sdk/proto/go/provider.pb.go (hunk 4 above)
Comment (swgillespie) [reply sdk/proto/go/provider.pb.go]: So despite the digest changing, these two gzipped payloads (the old and the new) actually extract to identical contents.

Comment (swgillespie) [reply sdk/proto/go/provider.pb.go]: So THIS is why Bazel downloads its own Go toolchain...

**Comment by swgillespie** [comment]
thanks!

**Pull request merged**